snuba/migrations/snuba_migrations/outcomes/0003_outcomes_add_category_and_quantity.py

```diff
@@ -0,0 +1,168 @@
+    Column("key_id", UInt(64)),
+    Column("timestamp", DateTime()),
+    Column("outcome", UInt(8)),
+    Column("reason", String()),
+    Column("times_seen", UInt(64)),
+]
+
+new_materialized_view_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("key_id", UInt(64)),
+    Column("timestamp", DateTime()),
+    Column("outcome", UInt(8)),
+    Column("reason", String()),
+    Column("category", UInt(8, Modifiers(nullable=True))),
+    Column("quantity", UInt(64, Modifiers(nullable=True))),
+    Column("times_seen", UInt(64)),
+]
+
+
+class Migration(migration.MultiStepMigration):
+    """
+    Adds quantity and category columns to outcomes. updates materialized view and hourly table to support
+    category as a new dimension and quantity as a new measure.
+    """
+
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            operations.AddColumn(
+                storage_set=StorageSetKey.OUTCOMES,
+                table_name="outcomes_raw_local",
+                column=Column("quantity", UInt(32, Modifiers(nullable=True))),
+                after=None,
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.OUTCOMES,
+                table_name="outcomes_raw_local",
+                column=Column("category", UInt(8, Modifiers(nullable=True))),
+                after=None,
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.OUTCOMES,
+                table_name="outcomes_hourly_local",
+                column=Column("quantity", UInt(64, Modifiers(nullable=True))),
+                after=None,
+            ),
+            operations.RunSql(
+                storage_set=StorageSetKey.OUTCOMES,
+                statement="""
+            ALTER TABLE outcomes_hourly_local ADD COLUMN IF NOT EXISTS category UInt8,
+            MODIFY ORDER BY (org_id, project_id, key_id, outcome, reason, timestamp, category);
+            """,
+            ),  # note: this migration is not reversable!
+            operations.DropTable(
+                storage_set=StorageSetKey.OUTCOMES,
+                table_name="outcomes_mv_hourly_local",
+            ),
+            operations.CreateMaterializedView(
+                storage_set=StorageSetKey.OUTCOMES,
+                view_name="outcomes_mv_hourly_local",
+                destination_table_name="outcomes_hourly_local",
+                columns=new_materialized_view_columns,
+                query=f"""
+                    SELECT
+                        org_id,
+                        project_id,
+                        ifNull(key_id, 0) AS key_id,
+                        toStartOfHour(timestamp) AS timestamp,
+                        outcome,
+                        ifNull(reason, 'none') AS reason,
+                        ifNull(category,{DataCategory.ERROR}) as category,
```

Comment:
category actually can't be nullable on the hourly table since it needs to be added to the sorting key, similarly to reason. All null values should be errors, so I think this will work

Reply:
(main discussion of this on notion, perhaps this will not work)